src/ja/server/scheduler/default_algorithm.py

```diff
@@ -17,16 +31,150 @@ def __init__(self,
+
+    def _free_machines(self, job: DatabaseJobEntry, next_machines: List[WorkMachine]) -> List[WorkMachine]:
+        if self._cost_cache[job.job.uid] <= self._cost_func.preempting_threshold:
+            return next_machines
+
+        usable_machines = list(filter(lambda m: m.uid not in self._reserved_machines.values(), next_machines))
+        if job.job.uid in self._reserved_machines:
+            usable_machines += [m for m in next_machines if m.uid == self._reserved_machines[job.job.uid]]
+
+        return usable_machines
+
+    def _schedule_nonpreemptive(self,
+                                job: DatabaseJobEntry,
+                                next_schedule: ServerDatabase.JobDistribution,
+                                next_machines: List[WorkMachine]) -> bool:
+        # Try to schedule without preemption
+        if job.job.status is JobStatus.PAUSED:
+            # Special case: check that we have enough RAM and CPU
+            need_allocation = get_allocation_for_job(job.job, JobStatus.RUNNING)
+            if not job.assigned_machine.resources.allocate(need_allocation, test_only=True):
+                return False
+            self._set_state(job.job, next_schedule, job.assigned_machine, JobStatus.RUNNING)
+            return True
+
+        usable_machines = self._free_machines(job, next_machines)
+        non_preemptive = self._non_preemptive_policy.assign_machine(job, next_schedule, usable_machines)
+        if non_preemptive:
+            self._set_state(job.job, next_schedule, non_preemptive[0], JobStatus.RUNNING)
+            self._reserved_machines.pop(job.job.uid, None)  # Make sure we do not hold the reserved machine any longer
+            return True
+        return False
+
+    def _schedule_blocking(self,
+                           job: DatabaseJobEntry,
+                           next_schedule: ServerDatabase.JobDistribution,
+                           next_machines: List[WorkMachine]) -> None:
+        result = self._blocking_policy.assign_machine(job, next_schedule, self._free_machines(job, next_machines))
+        if result:
+            self._reserved_machines[job.job.uid] = result[0].uid
+
+    def _schedule_preemptive(self,
+                             job: DatabaseJobEntry,
+                             next_schedule: ServerDatabase.JobDistribution,
+                             next_machines: List[WorkMachine]) -> bool:
+        preemptive = self._preemptive_policy.assign_machine(job, next_schedule, self._free_machines(job, next_machines))
+        if not preemptive:
+            return False
+
+        (machine, preempted_jobs) = preemptive
+        for preempt in preempted_jobs:
+            self._set_state(preempt, next_schedule, machine, JobStatus.PAUSED)
+        self._set_state(job.job, next_schedule, machine, JobStatus.RUNNING)
+        return True
+
+    def _deepcopy_args(self, current_schedule: ServerDatabase.JobDistribution, available_machines: List[WorkMachine]) \
```

Comment:
you can add *@ staticmethod* to this if you want

Reply:
ok, done.